components/style/values/specified/position.rs

```diff
@@ -486,3 +490,179 @@ impl From<GridAutoFlow> for u8 {
+                            return Err(());
+                        }
+                    }
+                    if let Some(index) = area_indices.get(token).cloned() {
+                        if areas[index].columns.start != column || areas[index].rows.end != row {
+                            return Err(());
+                        }
+                        areas[index].rows.end += 1;
+                        current_area_index = Some(index);
+                        continue;
+                    }
+                    let index = areas.len();
+                    areas.push(NamedArea {
+                        name: token.to_owned().into_boxed_str(),
+                        columns: column..(column + 1),
+                        rows: row..(row + 1),
+                    });
+                    assert!(area_indices.insert(token, index).is_none());
+                    current_area_index = Some(index);
+                }
+                if let Some(index) = current_area_index {
+                    if areas[index].columns.end != column + 1 {
+                        assert!(areas[index].rows.start != row);
+                        return Err(());
+                    }
+                }
+                if row == 1 {
+                    width = column;
+                } else if width != column {
+                    return Err(());
+                }
+            }
+        }
+        Ok(TemplateAreas {
+            areas: areas.into_boxed_slice(),
+            strings: strings.into_boxed_slice(),
+            width: width,
+        })
+    }
+}
+
+impl ToCss for TemplateAreas {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        for (i, string) in self.strings.iter().enumerate() {
+            if i != 0 {
+                dest.write_str(" ")?;
+            }
+            string.to_css(dest)?;
+        }
+        Ok(())
+    }
+}
+
+impl Parse for TemplateAreas {
+    fn parse<'i, 't>(
+        _context: &ParserContext,
+        input: &mut Parser<'i, 't>,
+    ) -> Result<Self, ParseError<'i>> {
+        let mut strings = vec![];
+        while let Ok(string) = input.try(|i| i.expect_string().map(|s| s.as_ref().into())) {
+            strings.push(string);
+        }
+
+        TemplateAreas::from_vec(strings)
+            .map_err(|()| input.new_custom_error(StyleParseErrorKind::UnspecifiedError))
+    }
+}
+
+trivial_to_computed_value!(TemplateAreas);
+
+#[cfg_attr(feature = "gecko", derive(MallocSizeOf))]
+#[derive(Clone, Debug, PartialEq)]
+/// Not associated with any particular grid item, but can
+/// be referenced from the grid-placement properties.
+pub struct NamedArea {
+    /// Name of the `named area`
+    pub name: Box<str>,
+    /// Rows of the `named area`
+    pub rows: Range<u32>,
+    /// Columns of the `named area`
+    pub columns: Range<u32>,
+}
+
+/// Tokenize the string into a list of the tokens,
+/// using longest-match semantics
+struct Tokenizer<'a>(&'a str);
```

Comment:
This name is super-generic, let's rename this to `TemplateAreasTokenizer` or something like it?

Reply:
Yes, it's better to name it with `TemplateAreas` prefix.